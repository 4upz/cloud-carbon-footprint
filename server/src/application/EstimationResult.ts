export interface EstimationResult {
  readonly timestamp: Date
  readonly estimates: ServiceEstimate[]
}

export interface ServiceEstimate {
  readonly serviceName: string
  readonly wattHours: number
  readonly co2e: number
}
