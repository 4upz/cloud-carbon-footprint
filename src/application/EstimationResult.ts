export interface EstimationResult {
  readonly timestamp: Date
  readonly estimates: ServiceEstimate[]
}

interface ServiceEstimate {
  readonly serviceName: string
  readonly wattHours: number
  readonly co2e: number
}
