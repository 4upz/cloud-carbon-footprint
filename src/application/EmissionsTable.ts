import { EstimationResult } from './EstimationResult'
import * as console from 'console'
import Table from 'cli-table'
import moment from 'moment'

interface Total {
  wattHours: number
  co2e: number
}

type Totals = { [key: string]: Total }

const initialTotals = () => ({
  ebs: {
    wattHours: 0,
    co2e: 0,
  },
})

const displayWattHours = (wattHours: number) => `${wattHours.toFixed(2)} Watts`
const displayCo2e = (co2e: number) => `${co2e.toFixed(6)} Kg CO2e`

export default function EmissionsTable(estimations: EstimationResult[]): string {
  const table = new Table({
    head: ['Date (UTC)', 'EBS Wattage', 'EBS CO2e Emissions'],
    colWidths: [20, 20, 30],
  })

  const grandTotals: Totals = initialTotals()

  estimations.forEach((estimationResult) => {
    const subTotals: Totals = initialTotals()

    estimationResult.estimates.forEach((estimate) => {
      grandTotals[estimate.serviceName].wattHours += estimate.wattHours
      subTotals[estimate.serviceName].wattHours += estimate.wattHours
      grandTotals[estimate.serviceName].co2e += estimate.co2e
      subTotals[estimate.serviceName].co2e += estimate.co2e
    })

    table.push([
      moment(estimationResult.timestamp).utc().format('YYYY-MM-DD'),
      displayWattHours(subTotals['ebs'].wattHours),
      displayCo2e(subTotals['ebs'].co2e),
    ])
  })

  table.push(['Total', displayWattHours(grandTotals['ebs'].wattHours), displayCo2e(grandTotals['ebs'].co2e)])

  return table.toString()
}
