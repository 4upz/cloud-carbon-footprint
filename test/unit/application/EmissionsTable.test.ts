import { EstimationResult } from '@application/EstimationResult'
import EmissionsTable from '@application/EmissionsTable'
import moment = require('moment')

describe('EmissionsTable', () => {
  it('prints out the given estimation results grouped by date', () => {
    const input: EstimationResult[] = [
      {
        timestamp: moment('2020-07-10').toDate(),
        estimates: [
          {
            serviceName: 'ebs',
            wattHours: 1,
            co2e: 1,
          },
          {
            serviceName: 's3',
            wattHours: 2,
            co2e: 2,
          },
          {
            serviceName: 'ec2',
            wattHours: 3,
            co2e: 3,
          },
          {
            serviceName: 'elasticache',
            wattHours: 4,
            co2e: 4,
          },
        ],
      },
      {
        timestamp: moment('2020-07-09').toDate(),
        estimates: [
          {
            serviceName: 'ebs',
            wattHours: 7,
            co2e: 8,
          },
          {
            serviceName: 's3',
            wattHours: 55,
            co2e: 1,
          },
          {
            serviceName: 'ec2',
            wattHours: 90,
            co2e: 77,
          },
          {
            serviceName: 'elasticache',
            wattHours: 747,
            co2e: 787,
          },
        ],
      },
    ]

    const result: string = EmissionsTable(input)

    expect(result).toMatchSnapshot()
  })
})
