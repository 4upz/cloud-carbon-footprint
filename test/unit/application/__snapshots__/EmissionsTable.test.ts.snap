// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmissionsTable prints out the given estimation results for the provided services 1`] = `
"| Date (UTC)   | EBS Wattage       | EBS CO2e Emissions     | S3 Wattage        | S3 CO2e Emissions      | ELASTICACHE Wattage| ELASTICACHE CO2e Emissions| SUM Wattage       | SUM CO2e Emissions     
| 2020-07-09   | 7.00 Watts        | 8.000000 Kg CO2e       | 55.00 Watts       | 1.000000 Kg CO2e       | 747.00 Watts      | 787.000000 Kg CO2e     | 899.00 Watts      | 873.000000 Kg CO2e     
| 2020-07-10   | 1.00 Watts        | 1.000000 Kg CO2e       | 2.00 Watts        | 2.000000 Kg CO2e       | 4.00 Watts        | 4.000000 Kg CO2e       | 10.00 Watts       | 10.000000 Kg CO2e      
| Total        | 8.00 Watts        | 9.000000 Kg CO2e       | 57.00 Watts       | 3.000000 Kg CO2e       | 751.00 Watts      | 791.000000 Kg CO2e     | 909.00 Watts      | 883.000000 Kg CO2e     "
`;

exports[`EmissionsTable prints out the given estimation results grouped by date 1`] = `
"| Date (UTC)   | EBS Wattage       | EBS CO2e Emissions     | S3 Wattage        | S3 CO2e Emissions      | EC2 Wattage       | EC2 CO2e Emissions     | ELASTICACHE Wattage| ELASTICACHE CO2e Emissions| SUM Wattage       | SUM CO2e Emissions     
| 2020-07-09   | 7.00 Watts        | 8.000000 Kg CO2e       | 55.00 Watts       | 1.000000 Kg CO2e       | 90.00 Watts       | 77.000000 Kg CO2e      | 747.00 Watts      | 787.000000 Kg CO2e     | 899.00 Watts      | 873.000000 Kg CO2e     
| 2020-07-10   | 1.00 Watts        | 1.000000 Kg CO2e       | 2.00 Watts        | 2.000000 Kg CO2e       | 3.00 Watts        | 3.000000 Kg CO2e       | 4.00 Watts        | 4.000000 Kg CO2e       | 10.00 Watts       | 10.000000 Kg CO2e      
| Total        | 8.00 Watts        | 9.000000 Kg CO2e       | 57.00 Watts       | 3.000000 Kg CO2e       | 93.00 Watts       | 80.000000 Kg CO2e      | 751.00 Watts      | 791.000000 Kg CO2e     | 909.00 Watts      | 883.000000 Kg CO2e     "
`;
